dm/syncer/validate_worker.go

```diff
@@ -0,0 +1,380 @@
+	}
+	colCnt := len(rows[0].data)
+	cond := &Cond{
+		Table:     table,
+		ColumnCnt: colCnt,
+		PkValues:  pkValues,
+	}
+	var failedRows map[string]*validateFailedRow
+	var err error
+	if deleteChange {
+		failedRows, err = vw.validateDeletedRows(cond)
+	} else {
+		failedRows, err = vw.validateInsertAndUpdateRows(rows, cond)
+	}
+	if err != nil {
+		vw.L.Warn("fail to validate row changes of table", zap.Error(err))
+		return nil, err
+	}
+	return failedRows, nil
+}
+
+func (vw *validateWorker) validateDeletedRows(cond *Cond) (map[string]*validateFailedRow, error) {
+	targetRows, err := vw.getTargetRows(cond)
+	if err != nil {
+		return nil, err
+	}
+
+	failedRows := make(map[string]*validateFailedRow, len(targetRows))
+	for key, val := range targetRows {
+		failedRows[key] = &validateFailedRow{tp: hasDeletedRow, dstData: val}
+	}
+	return failedRows, nil
+}
+
+func (vw *validateWorker) validateInsertAndUpdateRows(rows []*rowChange, cond *Cond) (map[string]*validateFailedRow, error) {
+	failedRows := make(map[string]*validateFailedRow)
+	sourceRows := getSourceRowsForCompare(rows)
+	targetRows, err := vw.getTargetRows(cond)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(targetRows) > len(sourceRows) {
+		// if this happens, downstream should have removed the primary key
+		vw.L.Debug("more data on downstream, may come from other client")
+	}
+
+	tableInfo := cond.Table.Info
+	for key, sourceRow := range sourceRows {
+		targetRow, ok := targetRows[key]
+		if !ok {
+			failedRows[key] = &validateFailedRow{tp: rowNotExist}
+			continue
+		}
+
+		eq, err := vw.compareData(sourceRow, targetRow, tableInfo.Columns[:cond.ColumnCnt])
+		if err != nil {
+			return nil, err
+		}
+		if !eq {
+			failedRows[key] = &validateFailedRow{tp: rowDifferent, dstData: targetRow}
+		}
+	}
+	return failedRows, nil
+}
+
+// a simplified version of https://github.com/pingcap/tidb-tools/blob/d9fdfa2f9040aab3fab7cd11774a82226f467fe7/sync_diff_inspector/utils/utils.go#L487-L606
+func (vw *validateWorker) compareData(sourceData, targetData []*sql.NullString, columns []*model.ColumnInfo) (bool, error) {
+	for i, column := range columns {
+		data1, data2 := sourceData[i], targetData[i]
+		if data1.Valid != data2.Valid {
+			return false, nil
+		}
+		str1, str2 := data1.String, data2.String
+		if str1 == str2 {
+			continue
+		} else if column.FieldType.Tp == tidbmysql.TypeFloat || column.FieldType.Tp == tidbmysql.TypeDouble {
+			// source and target data have different precision?
+			num1, err1 := strconv.ParseFloat(str1, 64)
+			num2, err2 := strconv.ParseFloat(str2, 64)
+			if err1 != nil || err2 != nil {
+				// should not happen
+				return false, errors.Errorf("convert %s, %s to float failed, err1: %v, err2: %v", str1, str2, err1, err2)
+			}
+			if math.Abs(num1-num2) <= 1e-6 {
+				continue
+			}
+		}
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (vw *validateWorker) getTargetRows(cond *Cond) (map[string][]*sql.NullString, error) {
```

Comment:
for deleted row type check, only selec pk is enough? maybe add another func to do this

Reply:
in next pr about checkpoint and pereist error rows(validate failed after N tries), need to save it's current data into a meta table, so user can view the detail